README.md
# dashc, client emulator for DASH video streaming

## Description

dashc is an emulator of the video player, it was intended to be used as a test tool. The original goal was to have a lightweight test tool for network performance experiments with a real DASH video streaming traffic.

dashc has several options which can be changed through the command line arguments:

* video adaptation algorithm (Conventional, BBA-0/BBA-1/BBA-2, ARBITER);
* initial buffer size (in segments);
* last segment index for play;
* log file and folder name (the log file can be on/off);
* maximum buffer size (in seconds);
* persistent connection (on/off);
* generation of the meta information for BBA-2 and ARBITER (file with segment sizes);

## Installation

dashc had been used in Ubuntu 16.04/16.10/17.04/17.10/18.04 with x64 architecture (continuously during 2017), in Ubuntu 16.04.3 with Raspberry Pi 2/3 (briefly during 2017), in macOS High Sierra 10.13 (briefly during 2018). Ubuntu 16.04/16.10/17.04/17.10 were used for tests when they were the newest available versions - dashc might work or might not work with them right now.

the current version of dashc has been tested with 18.04.02.

To install (./configure might ask for a sudo password to install opam, m4, git, bubblewrap and other dependencies (for Linux only)):

```bash
./configure
make
```

During ./configure, some questions might be asked:

Q: ## Where should it be installed ? [/usr/local/bin]

A: Accept it (just an enter key on a keyboard, not a "y" and then enter key).

Q: Do you want opam to modify ~/.profile? [N/y/f]

A: y

Q: A hook can be added to opam's init scripts to ensure that the shell remains in sync with the opam environment when they are loaded. Set that up? [y/N]

A: y

To run tests (not necessary):

```bash
make test
```

## Usage

dashc can be run as simple as:

```bash
./dashc.exe play http://www.cs.ucc.ie/~jq5/www_dataset_temp/x264_4sec/bbb_10min/DASH_Files/VOD/bbb_enc_10min_x264_dash.mpd
```

or (all available options):

```bash
./dashc.exe play http://127.0.0.1/bbb.mpd [-adapt conv] [-initb 2] [-maxbuf 60] [-persist true] [-turnlogon true] [-logname now] [-subfolder qwe]
```

Where all flags except the link to the MPD file are optional. Possible flags are:

* [-adapt], adaptation algorithm (conv; bba-0; bba-1; bba-2; arbiter);
* [-initb], initial buffer (in segments);
* [-lastsegmindex], last segment index for play;
* [-logname], name of the log file ("now" means usage of the current time stamp and it is used by default);
* [-maxbuf], maximum buffer size (in seconds);
* [-persist], persistent connection (true/false);
* [-segmentlist], this parameter is used to tell where to get meta data infromation (segment sizes, for example, for BBA-2 and ARBITER). Possible options are head (dashc will send head requests before starting to stream), local (file with segment size should be at the same folder with dashc), remote (file with segment size should be at the same location where target MPD file is stored) get segment sizes from  local - local segmentlist_%mpd_name%.txt file. The default option is head;
* [-subfolder], subfolder for the file;
* [-turnlogon], turn on logging to file;
* [-gensegmfile], generate segmentlist_%mpd_name%.txt file only (it will be rewritten if exists);
* [-help],  print this help text and exit, (alias: -?);

The detailed help can be found by running it with the -help flag:

```bash
./dashc.exe play -help
```

If you don't want to wait while "meta information" is being downloaded "Starting of downloading headers for representation %NUMBER%" (segment sizes for BBA-1/2 and ARBITER) you can use -gensegmfile flag to pre-download one, for example, by typing

```bash
./dashc.exe play http://www.cs.ucc.ie/~jq5/www_dataset_temp/x264_4sec/ed_10min/DASH_Files/VOD/ed_enc_10min_x264_dash.mpd -gensegmfile true
```

file "segmentlist_ed_enc_10min_x264_dash.mpd.txt" will be created in the current folder. Then you can run it like

```bash
./dashc.exe play http://www.cs.ucc.ie/~jq5/www_dataset_temp/x264_4sec/ed_10min/DASH_Files/VOD/ed_enc_10min_x264_dash.mpd -adapt arbiter -segmentlist local
```

The typical log file will look like the example below (here the data only for the first 30 segments is shown, ARBITER adaptation algorithm is used, bbb-264 4 seconds segment size clip is used):

```bash
Seg_#  Arr_time  Del_Time  Stall_Dur  Rep_Level  Del_Rate  Act_Rate  Byte_Size  Buff_Level
    1       810       810          0        232      1225       248     124131       4.000
    2      1554       744          0        232      1487       276     138452       8.000
    3      2283       728          0        232      1496       272     136272      11.272
    4      3696      1413          0        374      1591       562     281170      13.858
    5      4273       576          0        752      1884       271     135890      17.282
    6      5571      1298          0        752      1963       637     318725      19.983
    7     10756      5184          0       1060      1322      1713     856959      18.799
    8     15020      4263          0        752       823       877     438974      18.535
    9     16572      1551          0        560      1350       523     261860      20.983
   10     17832      1260          0        560      1212       381     190970      23.723
   11     19904      2072          0        752      1117       578     289478      25.651
   12     22951      3046          0        752      1032       786     393344      26.605
   13     26412      3461          0        752      1019       882     441029      27.143
   14     28438      2026          0        752      1536       778     389046      29.117
   15     29046       607          0        752      1732       263     131513      32.510
   16     29996       949          0        752      2105       499     249943      35.560
   17     32062      2066          0       1060      2034      1050     525440      37.494
   18     34956      2893          0       1060      1111       803     401864      38.601
   19     39174      4218          0       1060      1232      1299     649641      38.382
   20     44807      5631          0       1060      1560      2196    1098391      36.750
   21     46942      2134          0       1060      1357       723     361990      38.616
   22     48440      1498          0       1060      1071       401     200595      41.118
   23     51428      2987          0       1060      1055       788     394241      42.130
   24     53142      1713          0       1060      1036       443     221995      44.416
   25     55589      2446          0       1060      1185       724     362490      45.970
   26     57293      1704          0        752      1590       677     338960      48.265
   27     59909      2615          0        752      1200       784     392369      49.649
   28     63206      3297          0       1060      1254      1034     517165      50.352
   29     66206      2999          0       1060      1740      1305     652588      51.353
   30     70260      4053          0       1060      1822      1847     923626      51.299
```

The columns include next information:

* segment number starting from 1;
* arrival time of a segment in milliseconds;
* time spent for delivery of this segment in milliseconds;
* stall (freeze) duration in milliseconds;
* representation rate of downloaded segment in Kbit/s (taken from MPD file);
* delivery rate of the network in Kbit/s (segment size divided by time for delivery);
* the actual bitrate of this segment (segment size divided by the segment duration) in Kbit/s;
* segment size in bytes;
* buffer level after this segment was just downloaded, in seconds;

## Possible datasets for usage

iVID-Datasets for AVC (H.264) and HEVC (H.265): <https://www.ucc.ie/en/misl/research/datasets/ivid_dataset/>. The example of the log file above was taken with this dataset and caddy web server (<https://en.wikipedia.org/wiki/Caddy_(web_server)>).

AVC and HEVC UHD 4K DASH Datasets: <https://www.ucc.ie/en/misl/research/datasets/ivid_uhd_dataset/>


## Support

dashc is not in active development, if you have an issue, please, create a new one in github with necessary information (type of MPD file, OS version and etc.) - maybe somebody will answer.

## FAQ

Q: I have opam 1.2.2 already, do I really need to install opam 2?

A: Yes, read the <https://opam.ocaml.org/blog/opam-2-0-0-repo-upgrade-roadmap/>.

Q: X package (dependency) cannot be installed.

A: Update of the packages database might help. opam-repository (<https://github.com/ocaml/opam-repository/pulls>) changes constantly - some packages might break temporarily.

```bash
opam update          # Update the packages database
opam upgrade         # Bring everything to the latest version possible
```

Q: After the first ./configure, the make prints something like:

```bash
make: dune: Command not found
Makefile:4: recipe for target 'all' failed
make: *** [all] Error 127
```

A: Log out of the OS and log in back might help.

Q: Why one of Cohttp library files is swapped during ./configure?

A: Cohttp library has (had?) interface only for non-persistent connections (requests), so it is slightly modifed by this file. configure keeps the hash of the last commit (<https://github.com/mirage/ocaml-cohttp>) when everything was fine, it make sense to update it sometimes (cohttp might get support of a new OCaml compiler or new features/bug fixes).

Q: Why there was a need in a new implementation instead of updating already existing similar solutions?

A: Initially GPAC player was supposed to be used for doing the main phd research, but it didn't work properly for some reason. At the same time it was clear that making a custom small emulator of the video player won't take much time.

Q: Why OCaml was used (not a C/C++/Java/C#/Python)?

A: The main reason for OCaml choice was a past expirience with it. It took around two weeks of development to make a tool with conventional and BBA-0/1/2 algorithms during December 2016 (Christmas holidays!).

Q: What is OCaml?

A: A good set of resources can be found here:
* <https://ocaml.org/learn/>
* <https://github.com/rizo/awesome-ocaml>
* <https://github.com/vramana/awesome-reasonml>
* <https://dev.realworldocaml.org/>
* <http://www.cs.cornell.edu/courses/cs3110>
* <http://alan.petitepomme.net/cwn/> (OCaml weekly news)
* <https://discuss.ocaml.org/>
* Some videos about OCaml (<https://www.youtube.com/watch?v=v1CmGbOGb2I>, <https://www.youtube.com/watch?v=-J8YyfrSwTk>, <https://www.youtube.com/watch?v=FnBPECrSC7o>)

Q: Does anybody use OCaml in production?

A: <https://ocaml.org/learn/companies.html>

Q: Are there any good IDEs for OCaml?

A: Visual Studio Code with reasonml plugin  (<https://marketplace.visualstudio.com/items?itemName=freebroccolo.reasonml>) is good. Merlin and ocp-indent should be installed as well:
```bash
opam install merlin ocp-indent
```

Q: Are there any open source code examples in OCaml?

A: Some open source projects:
* <https://github.com/tezos>
* <http://www.cs.cornell.edu/courses/cs3110/archive/tournaments.html>
* <https://github.com/facebook/infer>
* <https://github.com/facebook/flow>
* <https://github.com/facebook/pyre-check>


## Citation

Aleksandr Reviakin, Ahmed H. Zahran, Cormac J. Sreenan. dashc : a highly scalable client emulator for DASH video. MMSys 2018 Open Dataset & Software Track, <https://cora.ucc.ie/bitstream/handle/10468/6408/6056_accepted_version_dashc_-_a_highly_scalable_client_emulator_for_DASH_video.pdf>

## Licence

This software has emanated from research conducted with the financial support of Science Foundation Ireland (SFI) under Grant Number 13/IA/1892.

University College Cork

GPL-2 License


CHANGES.md
Known issues
--------------------
* full_byte_range, main_byte_range, onDemand_byte_range profiles are not supported

0.1.22 (2019.04)
--------------------
* updates to configure and Makefile to remove build errors with updated verison of ocaml, cohttp, etc.

0.1.21 (2018.10)
--------------------
* Cohttp link was updated to 2018.10.19, version 1.2.0
* Update to OCaml 4.07.1 from 4.07.0

0.1.20 (2018.09)
--------------------
* content-length (from the header) is now used instead of the (String.length body)
  for segment size determination
* The bug related to double printing of the log on the screen was fixed.
  It happened when the logging to the file was turned off.
* Reconnection was added, if the server closes connection unexpectedly.
  It happens only once, the second time will force the tool to shutdown.
* mkdir is not called anymore with an empty string (no more "mkdir: missing operand" message)
* Ported from jbuilder to dune
* Cohttp link was updated to 2018.09.06, version 1.1.1
* Update to OCaml 4.07.0 from 4.06.1
* Parsing of mpd files is more generic now (#2, reported by pedosb)
* Ported from opam 1.2.2 to 2.0.0

0.1.19 (2018.04.30)
--------------------
* Cohttp link was updated to 2018.04.19, version 1.1.0
* Added parsing of the mediaPresentationDuration, lastsegmindex argument is optional now
* Fixed regression added in 0.1.18 (incorrect calculation of the total_number_of_repr)
* A few tests were added

0.1.18 (2018.03.02)
--------------------
* Update to OCaml 4.06.1 from 4.06.0
* Initial test was added (alcotest package is used)

0.1.17 (2018.01.07)
--------------------
* Necessary changes were applied for compatibility with core/async v0.10.0
  (v0.10.0 has support of glibc-2.26 which is used in Ubuntu 17.10)
* Cohttp master 2018.01.07, version 1.0.2 is used now instead of the master 2017.11.21, version 1.0.0

0.1.16 (2017.11.23)
--------------------
* now representations hash table is created according to the number of representations in the MPD file
* Added support of variable number of AdaptationSet tags in an MPD file
  (GPAC 0.7.2 creates several of them in comparison with GPAC 0.5.2 based on aspect ratio of video)

0.1.15 (2017.11.21)
--------------------
* Update cohttp from 0.99 to 1.0.0 (to be precise to the commit from 2017.11.21)
* Update to OCaml 4.06.0 from 4.05.0

0.1.14 (2017.11.20)
--------------------
* Fix in BBA-2: prev_time_for_delivery previously was calculated by dividing int by int (us / 1_000_000), what gave incorrect result (it was technically correct, but precision of 1 second is useless). Previously the previous representation level would be determined incorrectly based on an approximate representation rate. This changed was applied in BBA-1 (in calculate_reservoir as well), BBA-2, ARBITER
* full, live, main profiles (created by GPAC) are supported. They were supported before, but checked only now

0.1.13 (2017.09.29)
--------------------
* Fixes in ARBITER, Conventional algorithms and in print_result related to possible overflow of int in 32 bit OS.
It happens for sure in Raspberry Pi 3 in 32 bit OS. The 64 bit float will help to avoid it.

0.1.12 (2017.09.27)
--------------------
* Support of SegmentList type of MPD was added

0.1.11 (2017.09.26)
--------------------
* Fixes in throughput mean and throughput variance calculation in ARBITER:
previously results starting from the beginning were used for calculations instead of the most recent one

0.1.10 (2017.09.25)
--------------------
* Fix in BBA-1 (BBA-2 uses BBA-1 code) in selection of the next requested representation,
previously the representation level could've been changed only by one level up/down

0.1.9 (2017.08.24)
--------------------
* Added flag to change an implementation of conventional algorithm to an alternative one
* Port to Jbuilder

0.1.8 (2017.08.14)
--------------------
* Conventional algorithm coefficients were changed from 0.2 to 0.4 and from 0.8 to 0.6

0.1.7 (2017.07.25)
--------------------
* Update to OCaml 4.05.0 from 4.02.2 (4.02.2 was set by mistake instead of 4.04.2 in 0.1.5)

0.1.6 (2017.07.25)
--------------------
* Update to cohttp/cohttp-async 0.99

0.1.5 (2017.07.16)
--------------------
* default OCaml version is set back to 4.04.2
conduit and cohttp depend on ppx_deriving, ppx_deriving 4.1 has
Available	ocaml-version >= "4.02.1" & ocaml-version < "4.05" & opam-version >= "1.2"

0.1.4 (2017.07.15)
--------------------
* default OCaml version is set to 4.05.0
* change back to opam-based core library (in 0.9.3 there was an important for Raspberry Pi build fix)
https://github.com/janestreet/base/issues/15

0.1.3 (2017.06.07)
--------------------
* new repository from janestreet, mentioned in 0.1.2, was not added actually,
and it is not now, because these libraries were recently updated in opam, so, no need
* ocaml-cohttp folder added to .gitignore

0.1.2 (2017.05.03)
--------------------
* Update to the new janestreet repository:
opam repo add janestreet-dev https://ocaml.janestreet.com/opam-repository

0.1.1 (2017.04.26)
--------------------
* ARBITER now has 60 seconds maxbuf instead of taking it from command line parameters
* ARBITER was added somewhen in January 2017 (right after 0.1.0)

0.1.0 (2016.12.30)
--------------------
* Initial release
* implemented algorithms: Conventional (Probe and Adapt: Rate Adaptation for HTTP Video Streaming At Scale, https://arxiv.org/abs/1305.0510), bba-0, bba-1, bba-2 (A Buffer-Based Approach to Rate Adaptation: Evidence from a Large Video Streaming Service, http://yuba.stanford.edu/~nickm/papers/sigcomm2014-video.pdf)


src/dashc.ml
(*
 *	dashc, client emulator for DASH video streaming
 *	Copyright (c) 2016-2018, Aleksandr Reviakin, University College Cork
 *
 *	This program is free software; you can redistribute it and/or
 *	modify it under the terms of the GNU General Public License
 *	as published by the Free Software Foundation; either version 2
 *	of the License, or (at your option) any later version.
 *
 *	This program is distributed in the hope that it will be useful,
 *	but WITHOUT ANY WARRANTY; without even the implied warranty of
 *	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *	GNU General Public License for more details.
 *
 *	You should have received a copy of the GNU General Public License
 *	along with this program; if not, write to the Free Software
 *	Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
 *	02110-1301, USA.
*)

open Core

let command =
  Command.group ~summary:"Modes: play"
  [ "play", Playback.play ]

let () = Command.run command ~version:"0.1.21" ~build_info:"OCaml 4.07.1 was used"


src/playback.ml
(*
 *	dashc, client emulator for DASH video streaming
 *	Copyright (c) 2016-2018, Aleksandr Reviakin, University College Cork
 *
 *	This program is free software; you can redistribute it and/or
 *	modify it under the terms of the GNU General Public License
 *	as published by the Free Software Foundation; either version 2
 *	of the License, or (at your option) any later version.
 *
 *	This program is distributed in the hope that it will be useful,
 *	but WITHOUT ANY WARRANTY; without even the implied warranty of
 *	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *	GNU General Public License for more details.
 *
 *	You should have received a copy of the GNU General Public License
 *	along with this program; if not, write to the Free Software
 *	Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
 *	02110-1301, USA.
*)

open Core
open Async
open Cohttp_async

open Representation
open Segm_result
open Adapt_algo

let supported_algorithms = ["conv"; "bba-0"; "bba-1"; "bba-2"; "arbiter"]
let algorithms_with_required_chunk_sizes = ["bba-1"; "bba-2"; "arbiter"]
let log_columns = ["Seg_#"; "Arr_time"; "Del_Time"; "Stall_Dur";
"Rep_Level"; "Del_Rate"; "Act_Rate"; "Byte_Size"; "Buff_Level"]

let is_buffer_full buffer_size maxbuf segment_duration =
  buffer_size > (maxbuf -. float_of_int segment_duration)

let next_repr_number representations results last_segment_index = function
  | Conv ->
    Conv.next_representation_level ~representations:representations ~results:results
  | BBA_0 algo ->
    BBA_0.next_representation_level
    ~algo:algo ~representations:representations ~results:results
  | BBA_1 algo ->
    BBA_1.next_representation_level
    ~algo:algo ~representations:representations ~results:results last_segment_index
  | BBA_2 algo ->
    BBA_2.next_representation_level
    ~algo:algo ~representations:representations ~results:results last_segment_index
  | ARBITER algo -> ARBITER.next_representation_level
    ~algo:algo ~representations:representations ~results:results last_segment_index

let try_request_next_chunk ?conn link_to_next_chunk =
  try_with (fun () ->
    Client.get ?conn:conn (Uri.of_string link_to_next_chunk)
    >>| fun (resp, body) -> Some (resp, body)
  )
  >>| function
  | Ok resp_and_body -> resp_and_body
  | Error e -> begin
    match (String.is_substring (Exn.to_string e) ~substring:"Connection closed by remote host") with
    | true -> None
    | false -> failwith @@ Exn.to_string e
    end

let rec playback
    ?conn
    ?outc
    ~root_link
    ~representations
    ~results
    ~absolute_start_time
    ~buffer_size
    ~segment_number
    ~initb
    ~maxbuf
    ~adapt_method
    ~last_segment_index
    ~segment_duration
    ~link
    ~persist =
  match is_buffer_full buffer_size maxbuf segment_duration with
  | true ->
    Clock.after (Time.Span.of_sec 0.1) >>= fun () ->
    playback
      ?conn:conn
      ?outc:outc
      ~root_link:root_link
      ~representations:representations
      ~results:results
      ~absolute_start_time:absolute_start_time
      ~buffer_size:(buffer_size -. 0.1)
      ~segment_number:(segment_number)
      ~initb:initb
      ~maxbuf:maxbuf
      ~adapt_method:adapt_method
      ~last_segment_index:last_segment_index
      ~segment_duration:segment_duration
      ~link:link
      ~persist:persist
  | false ->
    let start = Time.now () in
    let next_repr_number =
      next_repr_number representations results last_segment_index adapt_method in
    let next_repr = Hashtbl.find_exn representations next_repr_number in
    let link_to_next_chunk =
      (root_link ^ (link_of_media next_repr.media segment_number)) in
    try_request_next_chunk ?conn:conn link_to_next_chunk
    >>= fun resp_and_body -> begin
    match resp_and_body with
    | None -> begin
        open_connection link persist >>= fun conn ->
        try_request_next_chunk ?conn:conn link_to_next_chunk
        >>= fun resp_and_body_2 ->
        match resp_and_body_2 with
        | None -> failwith "The second disconnect in a raw"
        | Some (resp, body) ->
          print_endline "New connection was opened";
          return (resp, body, conn)
      end
    | Some (resp, body) -> return (resp, body, conn)
    end
    >>= fun (resp, body, conn) ->
    body |> Cohttp_async.Body.drain >>= fun () ->
    let content_length = match resp |> Cohttp_async.Response.headers |> Cohttp.Header.get_content_range with
    | Some content_length -> Int64.to_int_exn content_length
    | None -> 0
    in
    let current_time = Time.now () in
    let time_diff = Time.diff current_time start in
    let diff = Time.Span.to_sec time_diff  in
    let initb = match buffer_size >= (initb *. float_of_int segment_duration) with
    | true -> 0.
    | false -> initb
    in
    let buf = match buffer_size > (initb *. float_of_int segment_duration) with
    | true -> begin
      match buffer_size <= diff with
      | true -> 0.
      | false -> buffer_size -. diff
      end
    | false -> buffer_size
    in
    let stall = match buffer_size > (initb *. float_of_int segment_duration)
      && buffer_size < diff with
    (* check this condition for startup phaze *)
    | true -> int_of_float @@ (diff -. buffer_size) *. 1000.
    | false -> 0
    in
    let results =
      {
        segment_number = segment_number;
        arrival_time =
          int_of_float @@ Time.Span.to_us (Time.diff current_time absolute_start_time);
        time_for_delivery = int_of_float @@ Time.Span.to_us time_diff;
        stall_dur = stall;
        representation_rate = next_repr.bandwidth;
        actual_representation_rate = 8 * content_length / segment_duration;
        segment_size = content_length;
        buffer_level_in_momentum = (buf +. float_of_int segment_duration);
        repr_level = next_repr_number;
      } :: results
    in
    print_result (List.hd_exn results) None;
    let () = match outc with
    | Some outc -> print_result (List.hd_exn results) (Some outc)
    | None -> ()
    in
    match segment_number >= last_segment_index with
    | true -> Deferred.unit
    | false ->
      playback
        ?conn:conn
        ?outc:outc
        ~root_link:root_link
        ~representations:representations
        ~results:results
        ~absolute_start_time:absolute_start_time
        ~buffer_size:(buf +. float_of_int segment_duration)
        ~segment_number:(segment_number + 1)
        ~initb:initb
        ~maxbuf:maxbuf
        ~adapt_method:adapt_method
        ~last_segment_index:last_segment_index
        ~segment_duration:segment_duration
        ~link:link
        ~persist:persist

let check_input_algorithm_existence input_alg =
  match List.exists supported_algorithms ~f:(fun x -> x = input_alg) with
  | true -> ()
  | false ->
    failwith "The chosen adaptation algorithm is not supported, please check help" 

let create_log_channel logname log_folder v_number r_number = function
  | true ->
    begin match logname, v_number with
    | "now", _ ->
      begin match log_folder with
      | "" ->
        Some (Out_channel.create @@
          Time.to_filename_string (Time.now ()) ~zone:Time.Zone.utc)
      | _ ->
        Some (Out_channel.create
          (log_folder ^ "/" ^  Time.to_filename_string (Time.now ()) ~zone:Time.Zone.utc))
      end
    | _, "-V99" ->
      begin match log_folder with
      | "" -> Some (Out_channel.create logname)
      | _ -> Some (Out_channel.create (log_folder ^ "/" ^ logname))
      end
    | _, _ ->
      begin match log_folder with
      | "" ->
        Some (Out_channel.create
          ("trace-" ^ Time.to_filename_string (Time.now ())
          ~zone:Time.Zone.utc ^ "-" ^ logname ^ v_number ^ r_number ^ ".res"))
      | _ ->
        Some (Out_channel.create
          (log_folder ^ "/" ^ "trace-" ^ Time.to_filename_string (Time.now ())
          ~zone:Time.Zone.utc ^ "-" ^ logname ^ v_number ^ r_number ^ ".res"))
      end
    end
  | false -> None

let check_requirement_for_chunk_sizes input_alg =
  List.exists algorithms_with_required_chunk_sizes ~f:(fun x -> x = input_alg)

let chunk_sizes
    adapt_alg
    segm_size_from_file
    link representations
    last_segment_index
    root_link
    segmlist_mpd
    conn =
  match check_requirement_for_chunk_sizes adapt_alg with
  | true ->
    let chunk_sizes_per_repr = match segm_size_from_file with
    | "local" ->
        read_segment_size_file
          ?remote_string:None
          ~link:link
          ~number_of_representations:(Hashtbl.length representations)
          ~last_segment_index:last_segment_index
    | "remote" ->
        let segm_list_remote_link =
          root_link ^ "/" ^ "segmentlist_" ^ segmlist_mpd ^ ".txt" in
        Client.get ?conn:conn (Uri.of_string segm_list_remote_link)
        >>= fun (_, body) ->
        body |> Cohttp_async.Body.to_string >>= fun body_string ->
        read_segment_size_file
          ~remote_string:body_string
          ~link:link
          ~number_of_representations:(Hashtbl.length representations)
          ~last_segment_index:last_segment_index
    | _ ->
        download_chunk_sizes_per_repr
          ?conn:None
          ~root_link:root_link
          ~representations:representations
          ~last_segment_index:last_segment_index
    in
    chunk_sizes_per_repr >>| fun chunk_sizes_per_repr ->
    let average_chunk_size_per_repr =
      calculate_average_chunk_size_per_repr chunk_sizes_per_repr in
    (chunk_sizes_per_repr, average_chunk_size_per_repr)
  | false -> return (Hashtbl.Poly.create ~size:10 (), [])

let new_adapt_method chunk_sizes_per_repr average_chunk_size_per_repr = function
  | "conv" -> return Conv
  | "bba-0" -> return @@ BBA_0 { BBA_0.maxb = 240. }
  | "bba-1" ->
      return @@ BBA_1 {
        BBA_1.maxb = 240.;
        BBA_1.chunk_sizes_per_repr = chunk_sizes_per_repr;
        average_chunk_size_per_repr = average_chunk_size_per_repr
      }
  | "bba-2" ->
      return @@ BBA_2 {
        BBA_1.maxb = 240.;
        BBA_1.chunk_sizes_per_repr = chunk_sizes_per_repr;
        average_chunk_size_per_repr = average_chunk_size_per_repr
      }
  | "arbiter" ->
      return @@ ARBITER {
        ARBITER.maxb = 60.;
        ARBITER.chunk_sizes_per_repr = chunk_sizes_per_repr;
      }
  | _ -> failwith "The chosen adaptation algorithm is not supported"

let print_log_header_on_screen =
  List.iteri log_columns
  ~f:(fun idx x -> print_string x;
    match not @@ phys_equal idx (List.length log_columns - 1) with
    | true -> print_string "  "
    | false -> print_endline ""
  )

let print_log_header_into_file = function
  | Some outc -> List.iteri log_columns
    ~f:(fun idx x -> Out_channel.output_string outc x;
      match not @@ phys_equal idx (List.length log_columns - 1) with
      | true -> Out_channel.output_string outc "  "
      | false -> Out_channel.newline outc
    );
  | None -> ()

let get_max_muf maxbuf = function
  | BBA_0 algo -> int_of_float algo.BBA_0.maxb
  | BBA_1 algo -> int_of_float algo.BBA_1.maxb
  | BBA_2 algo -> int_of_float algo.BBA_1.maxb
  | ARBITER algo -> int_of_float algo.ARBITER.maxb
  | _ -> maxbuf

let run_client
    ~link
    ~adapt_alg
    ~initb
    ~maxbuf
    ~persist
    ~turnlogon
    ~logname
    ~v_number
    ~r_number
    ~log_folder
    ~last_segment_index
    ~segm_size_from_file =
  try_with (fun () ->
      check_input_algorithm_existence adapt_alg;
      begin
      match (log_folder = "") with
      | true -> return ()
      | false -> Deferred.ignore @@ Sys.command @@ "mkdir -p " ^ log_folder
      end
      >>= fun _ ->
      let outc =
        create_log_channel
          logname log_folder ("-V" ^ v_number) ("-R" ^ r_number) turnlogon in
      open_connection link persist >>= fun conn ->
      Client.get ?conn:conn (Uri.of_string link) >>= fun (_, body) ->
      body |> Cohttp_async.Body.to_string >>= fun body ->
      let mpd = Xml.parse_string body in
      let representations : (int, representation) Hashtbl.t = repr_table_from_mpd mpd in
      let segment_duration = (Hashtbl.find_exn representations 1).segment_duration in
      let last_segment_index =
        get_last_segment_index mpd segment_duration last_segment_index in
      let root_link, segmlist_mpd = String.rsplit2_exn link ~on:'/' in
      let root_link = root_link ^ "/" in
      chunk_sizes
        adapt_alg
        segm_size_from_file
        link
        representations
        last_segment_index
        root_link
        segmlist_mpd conn
      >>= fun (chunk_sizes_per_repr, average_chunk_size_per_repr) ->
      new_adapt_method chunk_sizes_per_repr average_chunk_size_per_repr adapt_alg
      >>= fun adapt_method ->
      print_log_header_on_screen;
      print_log_header_into_file outc;
      let maxbuf = get_max_muf maxbuf adapt_method in
      playback
        ?conn:conn
        ?outc:outc
        ~root_link:root_link
        ~representations:representations
        ~results:[]
        ~absolute_start_time:(Time.now ())
        ~buffer_size:0.
        ~segment_number:1
        ~initb:(float_of_int initb)
        ~maxbuf:(float_of_int maxbuf)
        ~adapt_method:adapt_method
        ~last_segment_index:last_segment_index
        ~segment_duration:segment_duration
        ~link:link
        ~persist:persist
      >>= fun () ->
      match outc with
      | Some outc -> Out_channel.close outc; Deferred.unit
      | None -> Deferred.unit
  )
  >>| function
  | Ok () -> ()
  | Error e -> begin
    match (String.is_substring (Exn.to_string e) ~substring:"connection attempt timeout") with
      | true -> print_endline @@ "Connection attempt timeout (10 sec default) to " ^ link
      | false -> print_endline @@ Exn.to_string e
    end

let play =
  Command.async
    ~summary:"Dash simulation client"
    ~readme:(fun () -> "Usage example: ./dashc.native play http://10.0.0.1/bbb.mpd \
                        [-adapt conv] [-initb 2] [-maxbuf 60] [-persist true] \
                        [-turnlogon true] [-logname now] [-v 20] [-subfolder qwe]")
    (
      let open Command.Let_syntax in
      [%map_open
        let link = anon ("link_to_mpd" %: string)
        and adapt_alg = flag "-adapt" (optional_with_default "conv" string)
          ~doc:" adaptation algorithm [conv; bba-0; bba-1; bba-2; arbiter]"
        and initb = flag "-initb" (optional_with_default 2 int)
          ~doc:" initial buffer in segments"
        and maxbuf = flag "-maxbuf" (optional_with_default 60 int)
          ~doc:" maximum buffer size"
        and persist = flag "-persist" (optional_with_default true bool)
          ~doc:" persistent connection"
        and turnlogon = flag "-turnlogon" (optional_with_default false bool)
          ~doc:" turn on logging to file"
        and logname = flag "-logname" (optional_with_default "now" string)
          ~doc:" name of the log file (\"now\" means usage of the current time stamp \
                and it is used by default)"
        and v_number = flag "-v" (optional_with_default "99" string)
          ~doc:" video number according to StreamTraceAnalysis.py"
        and r_number = flag "-r" (optional_with_default "0" string)
          ~doc:" the run (R) number for the trace file"
        and subfolder = flag "-subfolder" (optional_with_default "" string)
          ~doc:" subfolder for the file"
        and last_segment_index = flag "-lastsegmindex" (optional int)
          ~doc:" last segment index for the playback"
        and gensegmfile = flag "-gensegmfile" (optional_with_default false bool)
          ~doc:" generate segmentlist_%mpd_name%.txt file only \
                (it will be rewritten if exists)"
        and segm_size_from_file =
          flag "-segmentlist" (optional_with_default "head" string)
          ~doc:" get segment sizes from \n \
                local - local segmentlist_%mpd_name%.txt file \n \
                remote - download remote segmentlist_%mpd_name%.txt file \
                  from the same folder where mpd is located \n \
                head - get segment size by sending head requests \
                  before playing for each segment per representation \
                usage of head requests works only with non-persistent connection so far \
                  and it it set by default (for this operation)"
        in
        fun () -> match gensegmfile with
        | true -> make_segment_size_file ~link:link ~persist:false
        | false -> run_client
            ~link:link
            ~adapt_alg:adapt_alg
            ~initb:initb
            ~maxbuf:maxbuf
            ~persist:persist
            ~turnlogon:turnlogon
            ~logname:logname
            ~v_number:v_number
            ~r_number:r_number
            ~log_folder:subfolder
            ~last_segment_index:last_segment_index
            ~segm_size_from_file:segm_size_from_file
      ]
    )